src/core/inductive.ml

```diff
@@ -55,17 +61,67 @@ let gen_ind_typ_codom : popt -> sym -> (tbox list -> tbox) -> string -> tbox =
     | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
   in aux [] !(ind_sym.sym_type)
 
+(** [prf_of p c ts t] returns a proof of the form
+    c.symb_prf ( (((p ts1) ...) tsn) t) where ts = [ts1;...;tsn]. *)
+let prf_of : tvar -> config -> tbox list -> tbox -> tbox = fun p c ts t ->
+  _Appl_symb c.symb_prf [_Appl (_Appl_list (_Vari p) ts) t]
+
+(** [preprocessing pos c ind_typ_list p_str x_str] computes elements of the
+    induction principles that we want to associate to a list of mutually
+    defined inductive types [ind_typ_list]. It returns a pair of lists:
+      - The first list is an association list mapping every inductive type to
+        a pair made of a predicate variable and its type.The name of a
+        predicate variable has the type tvar and its type has the type tbox.
+      - The second list is composed of the conclusions of the induction
+      principles.
+    The string [p_str] is used as prefix for predicate variables, and the
+    string [x_str] as prefix for type parameters. *)
+let preprocessing :
+      popt -> config -> sym list -> string -> string
+      -> ((sym * (tvar * tbox)) list * tbox list) =
+  fun pos c ind_typ_list p_str x_str ->
+  let prop = _Symb c.symb_Prop in
+
+  let rec aux :
+    int -> (sym * (tvar * tbox)) list -> tbox list -> sym list
+    -> ((sym * (tvar * tbox)) list * tbox list) =
+    fun i assoc_predicates conclusion_list ind_typ_list ->
+    match ind_typ_list with
+    | [] -> assoc_predicates, conclusion_list
+    | ind_sym::q ->
+        (* STEP 1 - Create the predicate (name + type) associated to the
+           symbol [ind_sym]. *)
+          (* A - Create the name of the predicate. *)
+        let p_str = p_str ^ string_of_int i in
+        let p = Bindlib.new_var mkfree p_str in
+          (* B - Create the type of the predicate p. *)
+        let codom ts = _Impl (_Appl_symb ind_sym ts) prop in
+        let p_type = gen_ind_typ_codom pos ind_sym codom p_str in
+        (* STEP 2 - Create the conclusion of the inductive principle
+           associated to [ind_sym]. *)
+        let codom ts =
+          let x = Bindlib.new_var mkfree x_str in
+          let t = Bindlib.bind_var x (prf_of p c ts (_Vari x)) in
+          _Prod (_Appl_symb ind_sym ts) t
+        in
+        let conclusion = gen_ind_typ_codom pos ind_sym codom p_str in
+        aux (i+1) ((ind_sym, (p, p_type))::assoc_predicates)
+          (conclusion::conclusion_list) q
+  in
+  aux 0 [] [] ind_typ_list
```

Comment:
This function in fact looks like a List.mapi function. Could you try to reformulate it as an instance of List.mapi please?